UefiCpuPkg/CpuDxeAArch64/CpuDxe.c

```diff
@@ -0,0 +1,179 @@
+    }
+
+    if (Resource->PhysicalStart == GicInterfaceBase) {
+      SkipGici = 1;
+    }
+
+    if (Resource->PhysicalStart == SerialRegBase) {
+      SkipSerial = 1;
+    }
+
+    Idx++;
+  }
+
+  if (SkipGici == 0) {
+    VirtualMemoryTable[Idx].PhysicalBase = GicInterfaceBase;
+    VirtualMemoryTable[Idx].VirtualBase  = VirtualMemoryTable[Idx].PhysicalBase;
+    VirtualMemoryTable[Idx].Length       = GIC_PAGES * EFI_PAGE_SIZE;
+    VirtualMemoryTable[Idx].Attributes   = ARM_MEMORY_REGION_ATTRIBUTE_DEVICE;
+    Idx++;
+  }
+
+  if (SkipGicd == 0) {
+    VirtualMemoryTable[Idx].PhysicalBase = GicDistributorBase;
+    VirtualMemoryTable[Idx].VirtualBase  = VirtualMemoryTable[Idx].PhysicalBase;
+    VirtualMemoryTable[Idx].Length       = GIC_PAGES * EFI_PAGE_SIZE;
+    VirtualMemoryTable[Idx].Attributes   = ARM_MEMORY_REGION_ATTRIBUTE_DEVICE;
+    Idx++;
+  }
+
+  if (SkipGicr == 0) {
+    VirtualMemoryTable[Idx].PhysicalBase = GicRedistributorBase;
+    VirtualMemoryTable[Idx].VirtualBase  = VirtualMemoryTable[Idx].PhysicalBase;
+    VirtualMemoryTable[Idx].Length       = GIC_PAGES * EFI_PAGE_SIZE;
+    VirtualMemoryTable[Idx].Attributes   = ARM_MEMORY_REGION_ATTRIBUTE_DEVICE;
+    Idx++;
+  }
+
+  if (SkipSerial == 0) {
+    VirtualMemoryTable[Idx].PhysicalBase = SerialRegBase;
+    VirtualMemoryTable[Idx].VirtualBase  = VirtualMemoryTable[Idx].PhysicalBase;
+    VirtualMemoryTable[Idx].Length       = EFI_PAGE_SIZE;
+    VirtualMemoryTable[Idx].Attributes   = ARM_MEMORY_REGION_ATTRIBUTE_DEVICE;
+    Idx++;
+  }
+}
+
+VOID
+InitMmu (
+  VOID
+  )
+{
+  VOID           *TranslationTableBase;
+  UINTN          TranslationTableSize;
+  RETURN_STATUS  Status;
+
+  ArmUplUpdateMemoryMap ();
+
+  Status = ArmConfigureMmu (
+             VirtualMemoryTable,
+             &TranslationTableBase,
+             &TranslationTableSize
+             );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Error: Failed to enable MMU\n"));
+  }
+}
+
+VOID
+ClearEioMode (
+  VOID
+  )
+{
+  UINTN  GicCtlr = ArmGicV3GetControlRegister ();
+
+  // According to < GICv3 and GICv4 Software Overview >
+  // Chapter 5.5 End of interrupt
+  //
+  // In the GICv3 architecture priority drop and deactivation
+  // can happen together and separately.
+  //
+  // Performs both the priority drop and deactivation.
+  if ((GicCtlr >> 1) & 0x1) {
+    GicCtlr &= 0xFFFFFFFFFFFFFFFDULL;
+    ArmGicV3SetControlRegister (GicCtlr);
+  }
+}
+
+/**
+  Initialize the state information for the CPU Architectural Protocol.
+
+  @param ImageHandle     Image handle this driver.
+  @param SystemTable     Pointer to the System Table.
+
+  @retval EFI_SUCCESS           Thread can be successfully created
+  @retval EFI_OUT_OF_RESOURCES  Cannot allocate protocol data structure
+  @retval EFI_DEVICE_ERROR      Cannot create the thread
+
+**/
+EFI_STATUS
+EFIAPI
+InitializeCpu (
+  IN EFI_HANDLE        ImageHandle,
+  IN EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  ClearEioMode ();

```

Comment:
Why is this needed here? What is special about UefiPayloadPkg that it is only needed here and not on other ARM platforms?

This belongs in the ArmGicDxe driver IMO.

Reply:
UefiPayload might loaded by bootloader or Linux, it might be running on top of Bare Metal or Virtualization environment. When it runs on top of Virtualization environment, EOI mode might be set (EOImode = 1) by Linuxboot, then we need to take "Priority drop" and "Deactivation" steps (two steps) when perform EOI (Described in chapter "5.5 End of interrupt", GiC V3 Spec).
And ArmGicDxe triggeres EOI for one step (treate EOImode=0), so we need to clear EOImode for UefiPayload.

Reply:
Actually, I don't think these belong in MdePkg at all. They are an implementation detail of the GIC driver, and belng wherever that driver lives.

I've pushed a series that cleans up the GIC and fixes this issue as well. This should remove the need for this hack, and also remove the need for any of the GIC-related definitions to be moved out of ArmPkg

https://github.com/tianocore/edk2/pull/10636/commits

Reply:
Great! I will rebase PR after https://github.com/tianocore/edk2/pull/10636/commits merged! Thanks for your information.